New package: SimpleGrad v0.1.0

JuliaRegistrator opened the pull request:
- Registering package: SimpleGrad
- Repository: https://github.com/mikesaint-antoine/SimpleGrad.jl
- Created by: @mikesaint-antoine
- Version: v0.1.0
- Commit: 64d4b7c9feb28c34dd64d1e171b60ed44cbad277
- Reviewed by: @mikesaint-antoine
- Reference: https://github.com/mikesaint-antoine/SimpleGrad.jl/issues/2
<!-- bf0c69308befbd3ccf2cc956ac8a4671113b158fd87a6d4e741b0173dab051a48f7dc43e593ffff4357bc647a3a83598ce5da40874506c00936906fdb8a38ec722f566cf76f2a2c770ca8fc633083163e20612aabd9c1b58793e371b7911f13dce90bc72f5942a9f2e822311e14bc63478d9f76bd6d75fb0a53022af68fd713abda4c452096d9f486c483e786812e24e53626d9ca1b463c5a3f99672d273cbaeb0c6e78910539927884f9cc1dadf4b92 -->

gdalle commented:
Hi and congrats on the package! If it is educational, does it really need to be registered?

JuliaRegistrator committed:
New package: SimpleGrad v0.1.0

UUID: 837eacb2-5a76-4cde-be3a-83fe353b65ae
Repo: https://github.com/mikesaint-antoine/SimpleGrad.jl.git
Tree: b6d0b38441fa6ac6a21f86e7a54f11a69536c2b6

Registrator tree SHA: 17aec322677d9b81cdd6b9b9236b09a3f1374c6a

mikesaint-antoine commented:
@gdalle Thanks! And my hope was for it to be a fully-usable package in addition to being educational. Ideally I'd like for users to be able to install it through Pkg and use it for real projects, but for it to also be simple enough that they can look under the hood and understand how it works even if they're beginners.

So I think it has the potential for real use-cases besides just being educational. In fact, I've used it recently at my day-job just for convenience, when I was trying to see if an unconventional problem could be reformulated and solved with gradient-descent. At the same time, it makes a tradeoff between performance and readability. So it's not as fast and comprehensive as Flux, for example, but it's more understandable and customizable.

gdalle commented:
> Ideally I'd like for users to be able to install it through Pkg and use it for real projects, but for it to also be simple enough that they can look under the hood and understand how it works even if they're beginners.

This is also possible for unregistered projects: your users can just do

```julia
pkg> add https://github.com/mikesaint-antoine/SimpleGrad.jl
```

My objection comes from the fact that the basic name makes it sound like some kind of reference package, even though it is not as polished as, say, https://github.com/tinygrad/tinygrad

goerz commented:
[noblock]

I would join @gdalle in raising the question whether an "educational package" like this needs to be registered. If the point of the package is that people should read the code or just play around with it, and make their own version of it while learning backward propagation, registering doesn't add very much.

Possible reasons to register it would be

(a) The package should be used as a dependency. This might even just be as a dependency for _projects_ (not packages), if this package is used is some larger educational context, like a course. If some larger group of students were to be expected to use it as a dependency for their class project, it might make things a lot easier if it's registered.

(b) The package is a very polished and authoritative resource. Registering it prevents any future use of the "SimpleGrad" name, so the package should live up to that responsibility. I would say this mainly means excellent and complete documentation, and a commitment that the package will be maintained in the longer term. I always feel better with nice package names like this if they are owned by an organization with multiple maintainers, rather than a personal account. That can be changed later-on too, though.

This package already looks pretty polished to me. I would ask that you publish the (existing) documentation on GitHub Pages, and link the documentation in the README and the about section. The README could probably use some fine-tuning, in any case.

But assuming the documentation gets spruced up a little bit, I wouldn't really stand in the way of registering the package. I'd basically put it up to @mikesaint-antoine to consider these points and to decide whether they want to go through with the registration.

P.S.: I started writing this before https://github.com/JuliaRegistries/General/pull/108692#issuecomment-2160693798 so it seems to me @mikesaint-antoine makes a good argument for registering it. Again, please polish the documentation a little more, but otherwise, I think this registration has my support.

goerz commented:
> even though it is not as polished as, say, https://github.com/tinygrad/tinygrad

I agree that for me to wholeheartedly endorse this package, It should be similarly well documented as tiny-grad. But based on what's already there, I feel like that might be achievable.

[noblock]

gdalle commented:
As a maintainer of the JuliaDiff organization, my main issue is with the name though. I'd be okay with something longer that makes the purpose clear, like SimplePedagogicGrad for instance

goerz commented:
[@goerz](https://github.com/JuliaRegistries/General/pull/108692#issuecomment-2160718585)
> I always feel better with nice package names like this if they are owned by an organization


[@gdalle](https://github.com/JuliaRegistries/General/pull/108692#issuecomment-2160724126)
> As a maintainer of the JuliaDiff organization

In my ideal universe, @mikesaint-antoine would polish this to the point where it is an exemplary educational resource, and the package would be moved to the JuliaDiff org, to ensure the possibility of long-term maintenance.

[noblock]

gdalle commented:
That would be cool indeed, although it does not guarantee that said maintenance actually occurs ^^

Still, I fail to see when one would want to use SimpleGrad in an actual project, especially now that DifferentiationInterface provides a no-hassle entry point to every other autodiff package

mikesaint-antoine commented:
Hi @gdalle and @goerz , thank you for the comments and feedback! A couple notes -- 

@gdalle 
> My objection comes from the fact that the basic name makes it sound like some kind of reference package, even though it is not as polished as, say, https://github.com/tinygrad/tinygrad

Tinygrad was actually the original inspiration for my project, and my ultimate goal is to continue building it out and adding functionality over time, similarly to how George Hotz started with a bare-bones tinygrad and built it out into a professional-quality library. This will definitely require a lot more polishing, as you noted, but my eventual hope is that SimpleGrad will be a reference package that will be used in real projects, while also preserving the simplicity and readability of the source code.

@gdalle
> Still, I fail to see when one would want to use SimpleGrad in an actual project, especially now that DifferentiationInterface provides a no-hassle entry point to every other autodiff package

I think the best argument I can make for real use-cases are that the readable and understandable code makes debugging and customizing easier. For example, last week I was working on a gradient descent application at my day-job and chose to use SimpleGrad rather than Flux, just because I was familiar with how the code worked and it make debugging much easier. Of course, this is a bit trivial to say — of course I’m more familiar with the code since I wrote it myself! But the idea is that with the educational website and simple source code, anybody could easily become familiar with SimpleGrad to the point of understanding how it works from scratch, making debugging much easier and troubleshooting much easier. So I think it could be a viable option for real-life use cases where the main priority is understandability, rather than speed/performance.

@goerz
> (a) The package should be used as a dependency. This might even just be as a dependency for _projects_ (not packages), if this package is used is some larger educational context, like a course. If some larger group of students were to be expected to use it as a dependency for their class project, it might make things a lot easier if it's registered.

Part of my goal is to develop a course around the package for my [educational Youtube channel](https://youtube.com/c/mikesaint-antoine) that I run in my free time. It’s a small channel but some of the more popular videos have 10k+ views, so it’s at least possible that there might be a couple thousand people installing the package as part of the course. So that’s one part of the educational goal, and the other part is the documentation that explains the source code from scratch.

@goerz
> I always feel better with nice package names like this if they are owned by an organization with multiple maintainers, rather than a personal account. That can be changed later-on too, though.

Part of the reason that I want to officially register my package is so that the name will be reserved. I’m a bit worried of a scenario where I develop unregistered package called SimpleGrad.jl and eventually have an educational website and Youtube course using that name, with potentially very many people familiar with it, but then somebody else officially registers a package called SimpleGrad and I need to stop using the name. Of course, I understand the point that reserving a simple name like that needs to be earned with a polished, high-quality package, and I am hoping to continue working on the package until it meets that standard.

@goerz
> In my ideal universe, [@mikesaint-antoine](https://github.com/mikesaint-antoine) would polish this to the point where it is an exemplary educational resource, and the package would be moved to the JuliaDiff org, to ensure the possibility of long-term maintenance.

Yes I would definitely be open to this if you both are open to it. I understand that the package requires more work, and I would like to continue working on it until it meets the standard required for registration, if possible. Thank you both for the comments!

goerz commented:
Based on all that, I would pretty strongly support registering this package as `SimpleGrad` (i.e., without a change of name).

My recommendation would be for

* @gdalle to invite @mikesaint-antoine to JuliaDiff
* Move `SimpleGrad` there
* Set up a proper documentation (i.e., GitHub action to build and deploy the existing documentation)
* Re-register from `JuliaDiff`

Or, if @gdalle doesn't like the idea of having this package in `JuliaDiff`, I would just keep going with this PR (and you can still update the documentation a bit within the three-day window).

> although it does not guarantee that said maintenance actually occurs 

Of course, but just I like having a bus-factor > 1 for packages with a "nice" name. Being in an established, org, at least if the original maintainer suddenly disappears, it's easy to appoint someone else if/when they are interested in taking over.

[noblock]

gdalle commented:
Thank you for the clarifications!

I now agree that registration under the name SimpleGrad.jl is a good thing, provided the package has a proper documentation. I have added "noblock" to all my comments, and I suggest you do the same @mikesaint-antoine :)

However I am not comfortable with adding someone to the JuliaDiff organization if I have never interacted with them in the Julia community. In due time, the package will possibly have a home in JuliaDiff, and so might its developer, but until then I would recommend keeping it on @mikesaint-antoine's personal GitHub profile. Given its importance for the YouTube channel and the project outlined above, I doubt it will go into zombie mode right away!

Pinging @oxinabox cause she might be interested

mikesaint-antoine commented:
@gdalle 
> However I am not comfortable with adding someone to the JuliaDiff organization if I have never interacted with them in the Julia community.

I think that's completely fair, especially since I'm basically a total beginner with Julia and package development.

@goerz 
> I would just keep going with this PR (and you can still update the documentation a bit within the three-day window).

Ok that sounds good to me. I'll try to keep working on the documentation over the next couple days to get it to the point where it's up to standard for registration. However, I also have a full time day-job so I can only really work on this in my free time, so it's possible that I might not have it ready in time and might need to resubmit in another couple weeks.

Also nice to meet you @oxinabox and please let me know if you have any feedback.

Thanks!

goerz commented:
[noblock] That sounds good to me! I agree that it seems that this package will be maintained at least for the foreseeable future, even under @mikesaint-antoine's personal account. If that ever changes, or if the package grows to a point where it should be highlighted in a larger context, it's always possible (and easy) to move it to an org later. Actually, there's precedent for packages to be hard-forked to an org when original maintainers have disappeared (I happen to be maintaining one of these myself: [DocumenterCitations](https://github.com/JuliaDocs/DocumenterCitations.jl), originally at [ali-ramadhan/DocumenterCitations](https://github.com/ali-ramadhan/DocumenterCitations.jl)). So I'm probably blowing this aspect out of proportion. Let's just keep going with the PR as-is.

> I'll try to keep working on the documentation over the next couple days to get it to the point where it's up to standard for registration. However, I also have a full time day-job so I can only really work on this in my free time, so it's possible that I might not have it ready in time and might need to resubmit in another couple weeks.

That's not necessary. The package already passes the (low) bar for a README/documentation that I would put on any other registration. Just maybe not quite the slightly higher bar that I might put on very general/nice/high-profile package names like this one. But the *promise* of adding more documentation in the near future is enough for me. This can be merged as-is, and then you can keep working on it afterwards.

(BTW, since people sometimes screw this up: You can add more commits during the three-day waiting period, but do not try to release a new version. Wait until this is merged.)

mikesaint-antoine commented:
@goerz 
> But the promise of adding more documentation in the near future is enough for me. This can be merged as-is, and then you can keep working on it afterwards.

Sounds good, thanks!

mikesaint-antoine commented:
@goerz @gdalle Just to clarify, is there anything else I need to do on my end at the moment (other than working on improving the documentation)? Or is it just a matter of waiting a couple days for the merge to go through now? 

And thank you both for your help and feedback!

[noblock]

goerz commented:
[noblock] You're good. Nothing to do except wait a couple of days until the auto-merge.

mikesaint-antoine commented:
@goerz Ok, thanks!

[noblock]

goerz commented:
[noblock] Actually, I think @gdalle still has to edit https://github.com/JuliaRegistries/General/pull/108692#issuecomment-2161138683 to include `[noblock]` 😀
